Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/calibration/ErrorHandler.java

```diff
@@ -0,0 +1,519 @@
+        for (int i = 0; i < arrayLength; i++) {
+            float[] values = Arrays.copyOfRange(pastForecastsFlattened, i * 3 * length, (i * 3 + 1) * length);
+            float[] upper = Arrays.copyOfRange(pastForecastsFlattened, (i * 3 + 1) * length, (i * 3 + 2) * length);
+            float[] lower = Arrays.copyOfRange(pastForecastsFlattened, (i * 3 + 2) * length, (i * 3 + 3) * length);
+            pastForecasts[i] = new RangeVector(values, upper, lower);
+
+        }
+        for (int i = arrayLength; i < forecastHorizon; i++) {
+            pastForecasts[i] = new RangeVector(length);
+        }
+        recomputeErrors(lastInputs, inputLength);
+    }
+
+    public void setUpperLimit(float[] upperLimit) {
+        if (upperLimit != null) {
+            checkArgument(upperLimit.length == this.upperLimit.length, "incorrect Length");
+            System.arraycopy(upperLimit, 0, this.upperLimit, 0, upperLimit.length);
+        }
+    }
+
+    public void setLowerLimit(float[] lowerLimit) {
+        if (lowerLimit != null) {
+            checkArgument(lowerLimit.length == this.lowerLimit.length, "incorrect Length");
+            for (int i = 0; i < lowerLimit.length; i++) {
+                checkArgument(lowerLimit[i] <= this.upperLimit[i], "lower limit is higher than upper limit");
+                this.lowerLimit[i] = lowerLimit[i];
+            }
+        }
+    }
+
+    /**
+     * updates the stored information (actuals) and recomputes the calibrations
+     * 
+     * @param input      the actual input
+     * @param deviations the deviations (post the current input)
+     */
+
+    public void updateActuals(double[] input, double[] deviations) {
+        int arrayLength = pastForecasts.length;
+        int inputLength = input.length;
+
+        for (int i = 0; i < lastInputs.length - inputLength; i++) {
+            lastInputs[i] = lastInputs[i + inputLength];
+        }
+        System.arraycopy(input, 0, lastInputs, lastInputs.length - inputLength, inputLength);
+
+        if (sequenceIndex > 0) {
+            // sequenceIndex indicates the first empty place for input
+            // note the predictions have already been stored
+            int inputIndex = (sequenceIndex + arrayLength - 1) % arrayLength;
+            float[] errorTuple = new float[lastInputs.length + 2 * inputLength + 2];
+            for (int y = 0; y < lastInputs.length; y++) {
+                errorTuple[y] = (float) lastInputs[y];
+            }
+            for (int i = 0; i < forecastHorizon; i++) {
+                if (sequenceIndex > i) {
+                    for (int j = 0; j < inputLength; j++) {
+                        RangeVector a = pastForecasts[inputIndex];
+                        int offset = i * inputLength;
+                        errorTuple[lastInputs.length] = i;
+                        errorTuple[lastInputs.length + 1] = forecastHorizon - i;
```

Comment:
Why do you need to store both i and forecastHorizon - i ?

Reply:
It seemed to set up a better constrained optimization -- basically it used the fact that RCF has to work harder to present diagonal lines ;)

Reply:
don't understand. Will ask offline.